john-deere-api.py

```diff
@@ -0,0 +1,122 @@
+import base64
+import datetime
+import json
+import uuid
+
+from flask import Flask, render_template, request, redirect
+import requests
+import urllib.parse
+
+app = Flask(__name__)
+
+settings = {
+    'clientId': '',
+    'clientSecret': '',
+    'wellKnown': 'https://signin.johndeere.com/oauth2/aus78tnlaysMraFhC1t7/.well-known/oauth-authorization-server',
+    'callbackUrl': 'http://localhost:9090/callback',
+    'scopes': 'ag1 ag2 ag3 eq1 eq2 org1 org2 files offline_access',
+    'state': uuid.uuid1(),
+    'idToken': '',
+    'accessToken': '',
+    'refreshToken': '',
+    'apiResponse': '',
+    'accessTokenDetails': '',
+    'exp': ''
+}
+
+
+def populate(data):
+    settings['clientId'] = data['clientId']
+    settings['clientSecret'] = data['clientSecret']
+    settings['wellKnown'] = data['wellKnown']
+    settings['callbackUrl'] = data['callbackUrl']
+    settings['scopes'] = data['scopes']
+    settings['state'] = data['state']
+
+
+def update_token_info(res):
+    # settings['idToken'] = res.json()['id_token']
+    token = res.json()['access_token']
+    settings['accessToken'] = token
+    settings['refreshToken'] = res.json()['refresh_token']
+    settings['exp'] = datetime.datetime.now() + datetime.timedelta(seconds=res.json()['expires_in'])
+    (header, payload, sig) = token.split('.')
+    payload += '=' * (-len(payload) % 4)
+    settings['accessTokenDetails'] = json.dumps(json.loads(base64.urlsafe_b64decode(payload).decode()), indent=4)
+
+
+def get_location_from_metadata(endpoint):
+    response = requests.get(settings['wellKnown'])
+    return response.json()[endpoint]
+
+
+def get_basic_auth_header():
+    return base64.b64encode(bytes(settings['clientId'] + ':' + settings['clientSecret'], 'utf-8'))
+
+
+def render_error(message):
+    return render_template('error.html', title='John Deere API with Python', error=message)
+
+
+@app.route("/", methods=['POST'])
+def start_oidc():
+    populate(request.form)
+    red = get_location_from_metadata('authorization_endpoint') + '?client_id=' + settings[
+        'clientId'] + '&response_type=code&scope=' \
+          + urllib.parse.quote(settings['scopes']) + '&redirect_uri=' + settings[
+        'callbackUrl'] + '&state=' + settings['state']
```

Comment:
Please improve the formatting here. Line breaks in the middle of array accessors make this hard to read. Also, please use descriptive variable names rather than the `red` abbreviation.